pkg/store/labelpb/label.go

```diff
@@ -39,15 +39,23 @@ func noAllocBytes(buf string) []byte {
 // ZLabelsFromPromLabels converts Prometheus labels to slice of labelpb.ZLabel in type unsafe manner.
 // It reuses the same memory. Caller should abort using passed labels.Labels.
 func ZLabelsFromPromLabels(lset labels.Labels) []ZLabel {
-	return *(*[]ZLabel)(unsafe.Pointer(&lset))
+	zlabels := make([]ZLabel, 0, lset.Len())
+	lset.Range(func(lbl labels.Label) {
+		zlabels = append(zlabels, ZLabel{Name: lbl.Name, Value: lbl.Value})
+	})
+	return zlabels
 }
 
 // ZLabelsToPromLabels convert slice of labelpb.ZLabel to Prometheus labels in type unsafe manner.
 // It reuses the same memory. Caller should abort using passed []ZLabel.
 // NOTE: Use with care. ZLabels holds memory from the whole protobuf unmarshal, so the returned
 // Prometheus Labels will hold this memory as well.
 func ZLabelsToPromLabels(lset []ZLabel) labels.Labels {
-	return *(*labels.Labels)(unsafe.Pointer(&lset))
```

Comment:
This is the main concern I have, we lose access to no alloc conversion between labels and zlabels. But this is the safest way to do it because the internal representation of labels changes and we cannot simply adopt it in thanos. That would lead to the wire-format changing in an incompatible way.

Reply:
Maybe you have some benchmark numbers of, for example, bucket store?

Reply:
I gave a try on this before and this point was exactly where i stopped.

For now, I think we should extract this functions to another file and have 2 implementations (one with the stringlabels enabled and one without) and keep the implementation when the `stringlabels` tag is disabled unchanged?

In the future, if the `stringlabels` is the way to go, we could change the [proto](https://github.com/thanos-io/thanos/blob/main/pkg/store/labelpb/types.proto) to match the `stringlabels` and come back with the no alloc conversion?

Reply:
I think it will be more complicated than that, because the downside of no allocation casts is that we adopt the same wire format as `labels.Labels`. If we have a no alloc ZLabel implementation for `stringlabels`, the wire format will change in a backwards incompatible way, so new old Thanos components will not understand new components.

So maybe we to have getter methods in the protobuf stubs, and this way we would either use the `stringlabels` format when accessing data, or use the old one with the labels slice. I think there are more changes coming to `labels.Labels` which include encoding strings using a symbols table: https://github.com/prometheus/prometheus/pulls.

Reply:
Oh wow, that's indeed a big change complexity. 

If we completely copy both slice and string we can just completely remove ZLabels, no point in them. Actually doing this now is quite nice idea as the benefit of stringlabels might make Zlabel removal still a benefit (!), which will be clear indication we don't need anything else.

However, I think I like @fpetkovski approach here. Currently ZLabels from/to still reuse strings (!), so there is still some benefit and includes same risk (reusing memory), and we need custom marshalling for it anyway. Just name is confusing (Zero alloc labels - Zlabels).

Reply:
>I think it will be more complicated than that, because the downside of no allocation casts is that we adopt the same wire format as labels.Labels. If we have a no alloc ZLabel implementation for stringlabels, the wire format will change in a backwards incompatible way, so new old Thanos components will not understand new components. This will also be problematic during rolling upgrades where instances are on different versions.

That make sense.. this would make the rollout problematic. But i still wonder if is not worth as a first step just go ahead with this and make sure we dont have any perf impact on the "main build".